### PR TITLE
overlord/state: fix lp1560628, in taskrunner let task goroutines clean up their tombs

### PR DESCRIPTION
**Pull request by pedronis**
Let the task goroutines clean their own tomb from the mapping, using Alive in Ensure made that happen too early when they hadn't set the task status yet causing spurious reruns.

Closes: #1560628

### CODE REVIEW

File: overlord/state/taskrunner.go [hunk 1]
```diff
@@ -184,6 +186,19 @@ func (r *TaskRunner) Ensure() {
...
 }
 
+// wait expectes to be called with th r.mu lock held
+func (r *TaskRunner) wait() {
+	for len(r.tombs) > 0 {
+		for id, t := range r.tombs {
+			r.mu.Unlock()
+			t.Wait()
+			r.mu.Lock()
+			delete(r.tombs, id)
```
Comment (niemeyer) [inline overlord/state/taskrunner.go]: By the time the t.Wait above returns, the task must have returned its function, and the task id shouldn't be there anymore per the changes above.

File: overlord/state/taskrunner.go [hunk 2]
```diff
@@ -193,10 +208,7 @@ func (r *TaskRunner) Stop() {
 		tb.Kill(nil)
 	}
 
-	for id, tb := range r.tombs {
-		tb.Wait()
-		delete(r.tombs, id)
-	}
+	r.wait()
```
Comment (niemeyer) [inline overlord/state/taskrunner.go]: I'm slightly concerned about that Ensure running in the background in those tests. That goroutine can freeze until the whole test is done, and trigger new goroutines to fire. We should probably have a stopped flag which says "Alright, at this point we have no background activity and we're not taking any further work on this runner!", so we bail when Ensure sees the flag.

Comment (pedronis) [inline overlord/state/taskrunner.go]: @niemeyer if I understand you mean a flag on the runner set in Stop() and obeyed in Ensure?

Comment (pedronis) [inline overlord/state/taskrunner.go]: done that

File: overlord/state/taskrunner.go (hunk 1 above)
Comment (pedronis) [inline overlord/state/taskrunner.go]: removed

**Comment by niemeyer** [comment]
Thanks, LGTM!

**Comment by mvo5** [comment]
:+1: And nice catch!

File: overlord/state/taskrunner_test.go [hunk 3]
```diff
@@ -114,7 +114,7 @@ func (ts *taskRunnerSuite) TestEnsureComplex(c *C) {
 
 	defer r.Stop()
 
-	// run in a loop to ensure ordering is correct by pure chance
+	// run in a loop to ensure ordering is not correct by pure chance
```
Comment (elopio) [inline overlord/state/taskrunner_test.go]: lol :D

**Pull request merged**